### ISSUE TITLE
Does damo support multiple "damo record"?

### ISSUE THREAD

**Issue by sj-aws** [damonitor/damo#3]
For continuing discussions from https://github.com/awslabs/damo/issues/52

**Comment by sjp38** [comment]
The original [question](https://github.com/awslabs/damo/issues/52#issue-1747231330) was,

> I have tried to run damo record in two different terminals together while the other is running, but see that the second damo record gets the following failure.
> [...]
> As I understand, /sys/kernel/mm/damon/admin/kdamonds/nr_kdamonds can be larger than 1 and it can create multiple kdamonds, but I'm wondering why this failure is shown in damo.
>
> I guess it might be a problem only in damo, but not from DAMON framework inside kernel.

And my answer was,

> In short, your assumption is correct. DAMON supports the use case, but damo's current official interface doesn't support the use case.  We're aiming to support such use case with the next generation of the interface, though. Please bear in mind with me until the next generation interface is ready.

Now users can spawn multiple kdamonds using `json` or `yaml` format DAMON parameters.  It doesn't support starting new kdamond while another kdamond is running, though.  For more details, please refer to [USAGE](https://github.com/damonitor/damo/blob/v2.4.8/USAGE.md#full-damon-parameters-update).

@honggyukim , would the current support of multiple kdamonds spawning resolve your issue?

**Comment by honggyukim** [comment]
The current way of spawning multiple kdamonds with `json` or `yaml` is fine for my use case.

> damo's current official interface doesn't support the use case. We're aiming to support such use case with the next generation of the interface, though. Please bear in mind with me until the next generation interface is ready.

I can just wait for the next generation interface as it's not urgent for me now.

**Comment by sjp38** [comment]
> I can just wait for the next generation interface as it's not urgent for me now.

Actually, the `json`/`yaml`-based multiple kdamonds spawning is the next generation interface that I mentioned at that time.  Maybe it was already supported at that time, but it was an unofficial support (not documented).

Do you think we still need a new interface?  What improvements compared to the current interface would be needed for that?

**Comment by honggyukim** [comment]
I think damo needs to support creating multiple kdamonds from command line even with multiple command executions.

But damo doesn't currently support such use cases. In hmsdk project, it generates json outputs twice, then concatenate them internally. I think it's not an easy interface for general users.

**Comment by sjp38** [comment]
I'm so sorry for this late response.  Somehow I missed your last comment.

> In hmsdk project, it generates json outputs twice, then concatenate them internally. I think it's not an easy interface for general users.

Thank you so much for the clarification.  I now understand the problem.

I want to keep the command line interface support only single kdamond, for simplicity.  I doubt if multiple kdamonds command line option is easy to maintain, and easy to be used by human.

What do you think about keeping `damo args damon` to support only single kdamond, but supporting the json/yaml output editing?  For example, it would be able to replace your manual conatenating part.  E.g.,

```
# damo args damon --damos_action migrate_hot > scheme.json
# damo args damon --damos_action migrate_cold --append scheme.json    # append the kdamond to those in scheme.json
# damo args damon --damos_action migrate_cold --remove scheme.json 0 # remove 0-th kdamond from those in scheme.json
```

**Comment by sjp38** [comment]
Btw, I think the original question of this issue is answered, and we're now discussing about ways to improve `damo` command line interface for multiple kdamonds.  What do you think about closing this issue and creating another one?

**Comment by sjp38** [comment]
@honggyukim I didn't get your answer to the above [question](https://github.com/damonitor/damo/issues/3#issuecomment-2466877494), but I think that features for multiple kdamonds could be useful in general.  Hence just implemented those as experimental feature.  The documentation is available at https://github.com/damonitor/damo/blob/a4120f90911abb6e3ef997191d6b94528afe1a9e/USAGE.md#multiple-kdamonds.  Please let me know if this solves your issue.

**Comment by sjp38** [comment]
Updated the subject of this issue to better describe what we are discussing now.

**Comment by sjp38** [comment]
TL; DR: Maybe we can just support multiple kdamonds from command line interface.

## Recap of the discussion.

Problem:
damo's DAMON parameters command line options does not support multiple kdamonds.  Users can pass `json` or `yaml` format DAMON parameters with `--kdamonds` option for multiple kdamonds usage.  But, the `json` or `yaml` format parmeters handling is not convenient enough.

Solution so far:
We worked for making `json` or `yaml` foramt DAMON parameters handling for multiple kdamonds easier to do with `damo args damon` command.  Also documented the usage.

## New thought.

As the command line options are being more advanced and better organized, now I think supporting multiple kdamonds from the command line interface might be somewhat doable.  The idea is, to introduce `--nr_ctxs` and `--nr_schemes` option, line `--damos_nr_filters`.  For example,

```
damo start \
    --numa_node 0 --monitoring_intervals_goal 4% 3 5ms 10s \
        --damos_action migrate_cold --damos_access_rate 0% 0% --damos_apply_interval 1s \
        --damos_quota_interval 1s --damos_quota_space 200MB --damos_quota_goal numa_mem_free_bp 0.5% 0 \
       --damos_filter reject young \
    --numa_node 1 --monitoring_intervals_goal 4% 3 5ms 10s \
        --damos_action migrate_hot --damos_access_rate 5% max --damos_apply_interval 1s \
        --damos_quota_interval 1s --damos_quota_space 200MB --damos_quota_goal numa_mem_util_bp 99.7% 0 \
       --damos_filter allow young \
    --nr_schemes 1 1 \
    --nr_ctxs 1 1
```

I will try to do this.

**Comment by honggyukim** [comment]
Hi @sjp38, it seems that the two kdamonds here is given using `--numa_node` as markers for each kdamond. But I don't clearly get what `--nr_schemes 1 1` and `--nr_ctxs 1 1` means here as we already know there is such restriction.

**Comment by sjp38** [comment]
Just pushed the initial implementation of the features by commit dcb56a2f3c41d8c67d721a7e877294c8215d8010.  Please read the commit message for example usage.

**Comment by sjp38** [comment]
More changes for multiple kdamonds control from command line, including addition of `--nr_targets` and documentation update are pushed now.  Please read 'Partial DAMON Parameters Update' and 'Partial DAMOS Parameters Update' sections of USAGE for the usage, and feel free to ask more questions.

Note that example command on https://github.com/damonitor/damo/issues/3#issuecomment-2764663341 is wrong.  Fixed version of it is like below.  Note that `--regions` options may better to be replaced with `--numa_node` or more accurate values.

```
./damo args damon --format report \
        --regions 100-200 --monitoring_intervals_goal 4% 3 5ms 10s \
        --damos_action migrate_cold 1 --damos_access_rate 0% 0% --damos_apply_interval 1s \
        --damos_quota_interval 1s --damos_quota_space 200MB --damos_quota_goal node_mem_free_bp 0.5% 0 \
        --damos_filter reject young \
        --regions 200-400 --monitoring_intervals_goal 4% 3 5ms 10s \
        --damos_action migrate_hot 0 --damos_access_rate 5% max --damos_apply_interval 1s \
        --damos_quota_interval 1s --damos_quota_space 200MB --damos_quota_goal node_mem_used_bp 99.7% 0 \
        --damos_filter allow young \
        --damos_nr_quota_goals 1 1 --damos_nr_filters 1 1 \
        --nr_targets 1 1 --nr_schemes 1 1 --nr_ctxs 1 1
kdamond 0
    context 0
        ops: paddr
        target 0
            region [100, 200) (100 B)
        intervals
            sample 5 ms, aggr 100 ms, update 1 s
            target 4 % accesses per 3 aggrs, [5 ms, 10 s] sampling interval
        nr_regions: [10, 1,000]
        scheme 0
            action: migrate_cold to node 1 per 1 s
            target access pattern
                sz: [0 B, max]
                nr_accesses: [0 %, 0 %]
                age: [0 ns, max]
            quotas
                0 ns / 200.000 MiB / 0 B per 1 s
                goal 0: metric node_mem_free_bp (nid 0) target 50 current 0
                priority: sz 0.1 %, nr_accesses 0.1 %, age 0.1 %
            watermarks
                metric none, interval 0 ns
                0 %, 0 %, 0 %
            filter 0
                reject young
kdamond 1
    context 0
        ops: paddr
        target 0
            region [200, 400) (200 B)
        intervals
            sample 5 ms, aggr 100 ms, update 1 s
            target 4 % accesses per 3 aggrs, [5 ms, 10 s] sampling interval
        nr_regions: [10, 1,000]
        scheme 0
            action: migrate_hot to node 0 per 1 s
            target access pattern
                sz: [0 B, max]
                nr_accesses: [5 %, 18,446,744,073,709,551,615 %]
                age: [0 ns, max]
            quotas
                0 ns / 200.000 MiB / 0 B per 1 s
                goal 0: metric node_mem_used_bp (nid 0) target 9,970 current 0
                priority: sz 0.1 %, nr_accesses 0.1 %, age 0.1 %
            watermarks
                metric none, interval 0 ns
                0 %, 0 %, 0 %
            filter 0
                allow young
```

**Comment by honggyukim** [comment]
Hi SeongJae,

> Hi [@honggyukim](https://github.com/honggyukim) having two `--numa_node` tells us it will have two DAMON contexts.

Ack.

> Today, each kdamond can have only one context, so we can also know the user is asking two kdamonds. But in future, DAMON might support multiple contexts per kdamond.

I knew this part.

> That's why we need `--nr_ctxs` here. Also, from the example command we know the user wants to create two schemes. But it is unclear to what context those schemes should be assigned. `--nr_schemes` is for that.

Hmm.. I'm not sure about this interface. It looks like the position of `--nr_ctxs` is the order of kdamond and the number given by `--nr_ctxs` are the number of contexts at the kdamond of the position. For example, `--nr_ctxs 3 1 5` means there are 3 kdamonds and each kdamond has 3, 1, and 5 contexts repectively. Is it correct?

> Just pushed the initial implementation of --nr_ctxs and --nr_schemes by commit https://github.com/damonitor/damo/commit/dcb56a2f3c41d8c67d721a7e877294c8215d8010.
> More changes for multiple kdamonds control from command line, including addition of --nr_targets

Likewise, `--nr_scheme` and `--nr_targets` works the same way. Could you confirm if it's correct? 

> Note that example command on https://github.com/damonitor/damo/issues/3#issuecomment-2764663341 is wrong. Fixed version of it is like below. Note that --regions options may better to be replaced with --numa_node or more accurate values.

Hmm.. It looks a bit confusing using `--numa_node` or `--regions` as a seperate marker for kdamond. So if parser sees those keywords, then does it create a new kdamond and use the followed up options?

**Comment by sjp38** [comment]
> For example, --nr_ctxs 3 1 5 means there are 3 kdamonds and each kdamond has 3, 1, and 5 contexts repectively. Is it correct?

That's correct.  Same for `--nr_schemes` and `--nr_targets`.

> Hmm.. It looks a bit confusing using --numa_node or --regions as a seperate marker for kdamond. So if parser sees those keywords, then does it create a new kdamond and use the followed up options?

`--numa_node` and `--regions` are not markers for kdamonds, since those are essential for each target rather than kdamond, and a single kdamond can have multiple targets.  But because there are `--nr_targets` and `--nr_ctxs`, `damo` can understand the users' intention in this case.

If this looks confusing, using `--ops` to mark each context explicitly could be a better option.

**Comment by sjp38** [comment]
The documentation is also updated: https://github.com/damonitor/damo/blob/next/USAGE.md#partial-damon-parameters-update

@honggyukim Coudl you please check if it solves this issue, or there is something to do more?

**Comment by sjp38** [comment]
@honggyukim Now we also have an example script showing the usage: https://github.com/damonitor/damo/blob/next/scripts/mem_tier.sh

Could you please confirm if this issue is resolved?

**Comment by sjp38** [comment]
Closing this issue for now, since apparently there is no update.  Please open a new one or comment here to ask re-opening of this if needed.

**Comment by honggyukim** [comment]
Hi @sjp38, I'm very sorry for the late response.

I just had a problem understanding the usage of multiple kdamonds with `--nr_targets`, etc as follows.
```
$ sudo ./damo args damon --format report \
    --ops paddr --regions 100-200 --damos_action migrate_cold 1 \
    --ops paddr --regions 400-700 --damos_action migrate_hot 0 \
    --nr_targets 1 1 --nr_schemes 1 1 --nr_ctxs 1 1
```
I'm fine closing this issue now as I now understand it how it works, but I hope if there is a more intuitive way.

Thanks very much for your patience and help.

**Comment by sjp38** [comment]
Starting from v3.4.1, damo will provide a new way to configure multiple kdamonds.  Instead of `nr_ctxs` and `nr_targets`, users can mark start of a new kdamond, context, target, scheme options with `--kdamond`, `--damon_ctx`, `--damon_target`, `--damos_scheme`, like below.

```
$ sudo ./damo args damon --format report \
    --kdamond --damon_ctx --ops paddr \
        --damon_target --regions 100-200 \
        --damon_target --regions 200-400 \
        --damos_scheme --damos_action migrate_cold 1 \
    --kdamond --damon_ctx --ops paddr \
		--damon_target --regions 400-700 \
        --damos_scheme --damos_action migrate_hot 0
kdamond 0
    context 0
        ops: paddr
        target 0
            region [100, 200) (100 B)
        target 1
            region [200, 400) (200 B)
        intervals: sample 5 ms, aggr 100 ms, update 1 s
        nr_regions: [10, 1,000]
        scheme 0
            action: migrate_cold to node 1 per aggr interval
            target access pattern
                sz: [0 B, max]
                nr_accesses: [0 %, 18,446,744,073,709,551,615 %]
                age: [0 ns, max]
            quotas
                0 ns / 0 B / 0 B per max
                priority: sz 0 %, nr_accesses 0 %, age 0 %
            watermarks
                metric none, interval 0 ns
                0 %, 0 %, 0 %
        access sample control
        enabled primitives: page_table
kdamond 1
    context 0
        ops: paddr
        target 0
            region [400, 700) (300 B)
        intervals: sample 5 ms, aggr 100 ms, update 1 s
        nr_regions: [10, 1,000]
        scheme 0
            action: migrate_hot to node 0 per aggr interval
            target access pattern
                sz: [0 B, max]
                nr_accesses: [0 %, 18,446,744,073,709,551,615 %]
                age: [0 ns, max]
            quotas
                0 ns / 0 B / 0 B per max
                priority: sz 0 %, nr_accesses 0 %, age 0 %
            watermarks
                metric none, interval 0 ns
                0 %, 0 %, 0 %
        access sample control
        enabled primitives: page_table
```

Hopefully this will be more intuitive and easy to use.
